v2/errors.go

```diff
@@ -5,19 +5,181 @@
 
 package v2
 
-// ErrContractInvalid is a specific error type for handling model validation failures. Type checking within
-// the calling application will facilitate more explicit error handling whereby it's clear that validation
-// has failed as opposed to something unexpected happening.
-type ErrContractInvalid struct {
-	errMsg string
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"runtime"
+)
+
+// ErrKind a categorical identifier used to give high-level insight as to the error type.
+type ErrKind string
+
+const (
+	// Constant Kind identifiers which can be used to label and group errors.
+	KindUnknown            ErrKind = "Unknown"
+	KindDatabaseError      ErrKind = "Database"
+	KindCommunicationError ErrKind = "Communication"
+	KindEntityDoesNotExist ErrKind = "NotFound"
+	KindContractInvalid    ErrKind = "ContractInvalid"
+	KindServerError        ErrKind = "UnexpectedServerError"
+	KindLimitExceeded      ErrKind = "LimitExceeded"
+	KindDuplicateName      ErrKind = "DuplicateName"
+	KindInvalidId          ErrKind = "InvalidId"
+	KindServiceUnavailable ErrKind = "ServiceUnavailable"
+)
+
+// EdgexError provides an abstraction for all internal EdgeX errors.
+// This exists so that we can use this type in our method signatures and return nil which will fit better with the way
+// the Go builtin errors are normally handled.
+type EdgexError interface {
+	// Error obtains the error message associated with the error.
+	Error() string
+	// DebugMessages returns a detailed string for debug purpose.
+	DebugMessages() string
+	// Message returns the first level error message without further details.
+	Message() string
+	// Code returns the status code of this error.
+	Code() int
+}
+
+// CommonEdgexError generalizes an error structure which can be used for any type of EdgeX error.
+type CommonEdgexError struct {
+	op string
```

Comment:
Add comment for what `op` is. Is there a more appropriate name then `op`?

Reply:
how about `callerInfo`?

Reply:
Like it.